src/info.rs

```diff
@@ -48,105 +51,73 @@ impl ConnectionInfo {
         Ref::map(req.extensions(), |e| e.get().unwrap())
     }
 
-    #[allow(clippy::cognitive_complexity, clippy::borrow_interior_mutable_const)]
     fn new(req: &RequestHead, cfg: &AppConfig) -> ConnectionInfo {
         let mut host = None;
         let mut scheme = None;
         let mut realip_remote_addr = None;
 
-        // load forwarded header
-        for hdr in req.headers.get_all(&header::FORWARDED) {
-            if let Ok(val) = hdr.to_str() {
-                for pair in val.split(';') {
-                    for el in pair.split(',') {
-                        let mut items = el.trim().splitn(2, '=');
-                        if let Some(name) = items.next() {
-                            if let Some(val) = items.next() {
-                                match &name.to_lowercase() as &str {
-                                    "for" => {
-                                        if realip_remote_addr.is_none() {
-                                            realip_remote_addr = Some(val.trim());
-                                        }
-                                    }
-                                    "proto" => {
-                                        if scheme.is_none() {
-                                            scheme = Some(val.trim());
-                                        }
-                                    }
-                                    "host" => {
-                                        if host.is_none() {
-                                            host = Some(val.trim());
-                                        }
-                                    }
-                                    _ => {}
-                                }
-                            }
-                        }
+        for el in req
+            .headers
+            .get_all(&header::FORWARDED)
+            .into_iter()
+            .filter_map(|hdr| hdr.to_str().ok())
+            .flat_map(|val| val.split(';'))
+            .flat_map(|pair| pair.split(','))
+        {
+            let mut items = el.trim().splitn(2, '=');
+            if let (Some(name), Some(val)) = (items.next(), items.next()) {
+                match name.to_lowercase().as_ref() {
+                    "for" => {
+                        realip_remote_addr.get_or_insert_with(|| val.trim());
                     }
+                    "proto" => {
+                        scheme.get_or_insert_with(|| val.trim());
+                    }
+                    "host" => {
+                        host.get_or_insert_with(|| val.trim());
+                    }
+                    _ => {}
                 }
             }
         }
```

Comment:
```suggestion
        req
            .headers
            .get_all(&header::FORWARDED)
            .into_iter()
            .filter_map(|hdr| hdr.to_str().ok())
            .flat_map(|val| val.split(';'))
            .flat_map(|pair| pair.split(','))
            .filter_map(|pair|
                let mut items = pair.trim().splitn(2, '=');
                Some((items.next()?, items.next()?))
            })
            .for_each(|(name, val)| {
                match name.to_lowercase().as_ref() {
                    "for" => &mut realip_remote_addr,
                    "proto" => &mut scheme,
                    "host" => &mut host,
                    _ => &mut Some(""),
                }
                .get_or_insert_with(|| val.trim());
            });
```

Reply:
for more functional style

Reply:
`for` loops are generally preferred over `for_each`, and that last match won't compile because of `_ => &mut Some(""),`. It's less about function style and more about maintainability.

Reply:
Oh, That really doesn't compile. I though the temporary would last until the first semicolon `;` - forgot about `}`.  But I still see `filter_map` would be nicer here ;-)

Reply:
i've hijacked this PR and incorporated the extra filter_map